### PR TITLE
Make lockTimeout more friendly for JPA map storage

### LINKED ISSUE #16616: Make lockTimeout better configurable in JpaMapStorageProviderFactory
#15890 Introduced a new config option for `JpaMapStorageProviderFactory` that configures how long the execution should wait for acquiring the pessimistic lock.

However, there was no default value configured so it can result in a deadlock. As part of this issue, we should investigate what is the best way to configure this. We should have some default value, but we also need to have the possibility to disable this as it is possible to configure global timeout on the database level. 

We should add this configuration also to Quarkus config.

In model tests, we set the `lockTimeout` value to 1000ms which may be low for some tests and can result in instabilities. We need to add some way to configure this per test case. We will need something like this:
1. Configure `lockTimeout` to some env property
2. At the beginning of the test store the original property value and replace it with some custom value
3. Reinitialize Keycloak factory
4. Do the test
5. Reset property to the original value
6. Reinitialize again

### PR DESCRIPTION
**Pull request by mhajas**
Closes #16616

### CODE REVIEW

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/transaction/StorageTransactionTest.java [hunk 1]
```diff
@@ -146,6 +152,10 @@ public void testLockObjectForModification() {
         tx3.begin();
         tx3.runStep(session -> LockObjectsForModification.lockRealmsForModification(session, () -> session.realms().getRealm(realmId)));
         tx3.commit();
+
+        if (originalTimeoutValue == null) System.clearProperty(LOCK_TIMEOUT_SYSTEM_PROPERTY);
+        else System.setProperty(LOCK_TIMEOUT_SYSTEM_PROPERTY, originalTimeoutValue);
+        reinitializeKeycloakSessionFactory();
```
Comment (ahus1) [inline testsuite/model/src/test/java/org/keycloak/testsuite/model/transaction/StorageTransactionTest.java:158]: When the test fails, the system property might not be cleaned up. 

Therefore wrap it in a try/finally, so it is cleaned up in all cases. 

In addition: the if/else should be wrapped with curly braces as best practice.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank you for this PR, @mhajas.

I see that it don't show the warning all the time any more, and only once on the initialization of the factory. The documentation for the property has also been generated in the server guide. The CLI option works as expected when running Quarkus 

When testing this with PostgreSQL and Quarkus, I saw that this worked. Still, the code `createEntityManagerFactory()` is overwritten in Quarkus and will never be called. Due to that, the warning is never shown. This is a bit more complex to handle AFAIK, so I suggest to create a new issue for that if you think this should be fixed. As there are no other databases as PostgreSQL and CockroachDB for the JPA map storage, this is not something we would need to fix now.

Find below some small comments that IMHO should be changed before this issue can be merged. Thanks!

**Comment by mhajas** [comment]
thank you for the review @ahus1. 
> When testing this with PostgreSQL and Quarkus, I saw that this worked. Still, the code createEntityManagerFactory() is overwritten in Quarkus and will never be called. Due to that, the warning is never shown. This is a bit more complex to handle AFAIK, so I suggest to create a new issue for that if you think this should be fixed. As there are no other databases as PostgreSQL and CockroachDB for the JPA map storage, this is not something we would need to fix now.

How are other options configured with Quarkus? For example `datasource`, `user`, etc.
https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java#L380-L392

If there is some custom config, is it possible to configure `javax.persistence.lock.timeout` there?

File: testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/JpaMapStorageCockroachdb.java [hunk 2]
```diff
@@ -97,7 +99,7 @@ public void updateConfig(Config cf) {
                 .config("password", COCKROACHDB_DB_PASSWORD)
                 .config("driver", "org.postgresql.Driver")
                 .config("driverDialect", "org.keycloak.models.map.storage.jpa.hibernate.dialect.JsonbPostgreSQL95Dialect")
-                .config("lockTimeout", "1000");
+                .config("lockTimeout", "${" + LOCK_TIMEOUT_SYSTEM_PROPERTY + ":}");
```
[the remark this answers is not in the document]
Comment (mhajas) [reply testsuite/model/src/test/java/org/keycloak/testsuite/model/parameters/JpaMapStorageCockroachdb.java:102]: I added the comment to the property so it is accessible everywhere the property is used. Is it ok?

**Review by ahus1 — DISMISSED** [review]
Thank you for updating this PR, ready to be merged!

**Comment by ahus1** [comment]
> How are other options configured with Quarkus

I suppose the settings would need to be picked from the configuration here https://github.com/keycloak/keycloak/blob/17bf092da638d03645c80725196284ffd91bffa5/quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java#L248 or here https://github.com/keycloak/keycloak/blob/17bf092da638d03645c80725196284ffd91bffa5/quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java#L297

From what I understand this is all done before the provider is instantiated, so needs to be duplicated somehow.

**Review by ahus1 — DISMISSED** [review]
Hi @mhajas - those changes look good, and the additional option  now ends up in Quarkus as expected. 

As you pushed a second commit for this PR, either squash those commits now, on when merging this PR. Thanks.

File: quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java [hunk 3]
```diff
@@ -322,6 +322,11 @@ private void configureDefaultPersistenceUnitProperties(ParsedPersistenceXmlDescr
             unitProperties.setProperty(AvailableSettings.JPA_TRANSACTION_TYPE, PersistenceUnitTransactionType.JTA.name());
         }
 
+        ConfigValue lockTimeoutConfigValue = getConfig().getConfigValue("kc.spi-map-storage-jpa-lock-timeout");
+        if (lockTimeoutConfigValue != null && lockTimeoutConfigValue.getValue() != null) {
+            unitProperties.setProperty(AvailableSettings.JPA_LOCK_TIMEOUT, lockTimeoutConfigValue.getValue());
+        }
+
```
Comment (vmuzikar) [inline quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java:329]: Reviewed just the Quarkus side and LGTM.

**Review by vmuzikar — APPROVED** [review]

**Review by ahus1 — APPROVED** [review]
LGTM!

**Comment by ahus1** [comment]
@hmlnarik - reviewed, ready to be merged. Thanks!

**Pull request merged**